### DISCUSSION TITLE
Disjunctive Facets Distribution

### DISCUSSION THREAD

**Discussion by gmourier** [meilisearch/product#187] [Feedback & Feature Proposal]
Today, MeiliSearch performs a single query that calculates the distribution of facets in a generalized way. We believe that this is not the right way to proceed. 

By making a single query the facet distribution is calculated for all filters, which is not ideal. When multiple facets are used in a search, the distribution of a facet should be computed from the other selected facets. Each facet must calculate its distribution independently.

<img width="455" alt="Capture d’écran 2021-08-25 à 17 56 24" src="https://user-images.githubusercontent.com/3692335/130824271-fd8c097f-0cb1-4c0c-b96a-fea23d7042a7.png">

Given this example, the distribution of `size` must be done by emitting a search query with the selected facet values for `resolution` as filter. The distribution of `resolution` must be done by filtering with the selected facet values for `size` and so on.. 

We have several possibilities in mind to change that.

## Ideas

Handle a multiple request array payload on the server side sent by a client.
Handle multiple concurrent requests on the frontend side (instant-meilisearch, sdks).
Handle it at the server-side without receiving a requests array but by exploring a way to compute and generate the disjunctive distribution of facets in a single query from the search payload. This way, the client code can be very simple.

**Comment by sietzekeuning** [comment]
It would be great if this could be done in 1 query

**Reply by patriq** [reply]
If done in 1 requests, you could ultimatly save bandwidth by having the facets distribution block the user from doing a "no results" request. If one needs to send multiple concurrent from the frontend side to achieve disjunctive facets distribution, this saved bandwidth goes to waste while potentially slowing down the application, since it now needs to wait for every concurrent request to finish before updating the UI.

I have seen it suggested multiple times that Algolia performed multiple requests in the past to achieve this (see [#945](https://github.com/meilisearch/MeiliSearch/issues/945#issuecomment-685526678)), but I don't think that's the case anymore, at least looking at their [E-Commerce demo](https://react-instantsearch.netlify.app/examples/e-commerce/search/), the frontend only does one request per filter change.

Although I am interested in the user experience, I am not using faceted search at the moment because it doesn't support disjuctive facets distribution without overcomplicating the client code, leading to more requests than what I was trying to prevent in the first place.

**Reply by gmourier** [reply]
Absolutely @patriq! 

In fact, Algolia does not make multiple requests. It's just that the client code sends an array of search `requests` to execute each facet disjointly. Maybe we could find a solution on the server-side so that we don't need to send an array of requests and execute the disjoint distribution automatically.

In both cases, it's only one network communication but it could be nice for people who don't use instant-search not to have to generate an array of `requests` for a search request that may need to distribute disjunctive facets. It could be much more convenient.

**Comment by dorinniscu** [comment]
Hello @gmourier Is there any plan for this feature to be implemented? Thanks

**Reply by gmourier** [reply]
Hello @dorinniscu 👋

It is not planned at this time. A solution to this problem will be explored when we will introduce the ability to execute multiple search requests in a single HTTP call.

**Reply by martin-ro** [reply]
Hi @gmourier, I've been browsing a couple of issues across different Meilisearch repo's about "disjunctive facets".

I would like to make sure that if that feature will be implemented at some point in the future this would enable 'OR' behavior across facets? Like it is implemented [here](https://codesandbox.io/s/ms-is-mese9?fontsize=14&hidenavigation=1&theme=dark)?

I couldn't find any issue regarding "multiple search requests in a single HTTP call" so it's hard to estimate when this feature would land. Is there a guesstimate you can give?

Thanks!

**Reply by jacobgahn** [reply]
Adding my vote that this is a big deal for our search integration. Disjunctive is the default behavior for refinementlists, but in that case the facet list is not a good experience because it does not show the full list of options.

**Comment by ar-siddiqui** [comment]
Bumping up, we need it too.

In our use case, we do not like the meilisearch API not returning 0 facets. We want to show users to see 0 value facets as well so that they can select them. The hack provided by this [PR](https://github.com/meilisearch/instant-meilisearch/pull/695) only works when you first do a no filter search, if you directly use a URL with specific rules it would not showcase them as the client does not have anything to remember from.

IMO the meilisearch API should always return 0 facets (or be configurable to return 0 facets) and the client could be configured to either display them or not.

For example, displaying 0 facets allow a user to add `spreadsheet` as well to the search
![image](https://user-images.githubusercontent.com/53625184/159548177-fb5892da-2730-4548-8c61-55394af5d28c.png)

But without 0 facets, the user is not even aware that a type `spreadsheet` exists.
![image](https://user-images.githubusercontent.com/53625184/159548495-135711aa-ddd7-43d3-b9f1-9fc7f7c1f358.png)

**Reply by jacobgahn** [reply]
Thank you! This will be a useful work around for us. However, I think the pattern that meilisearch should adopt is well defined by the OP. Really hoping to see parity with Algolia on this.

**Comment by sheldonj** [comment]
Add my vote to this as well. Is there any action on this, or a path that's defined where we could help push this forward?

**Comment by jeremylynch** [comment]
Is there any update on this one? This is a critical issue for our team.

**Reply by gmourier** [reply]
We'll try to fix this problem from the instant-meilisearch side first; I've just discussed it with @bidoubiwa; even if it's not perfect for a first iteration, it should allow instant-meilisearch users to be able to have the desired behavior. 

We will keep everyone informed about the outcome of that direction in this discussion.

I know that this may not be perfect but the core team is currently dealing with blockage points that can't be solved by any workaround.

Thanks for the feedback!

**Reply by benjick** [reply]
@gmourier great (for me) that this is coming for instant-meilisearch! Is there a timeframe for this?

**Reply by gmourier** [reply]
Hello @benjick, 

It should be released in early April for v1.1 (3rd April is the planned date)

**Comment by gmourier** [comment]
Hey everyone 👋

We have added disjunctive facets for instant-meilisearch in a beta version [here](https://github.com/meilisearch/instant-meilisearch/issues/884#issuecomment-1372126906), we are looking for your feedback. ✨

**Reply by MoustafaElsaghier** [reply]
Hey, @gmourier that's good news for us to have **disjunctive** feature in meilisearch, can you tell me is possible to be done through API too, or not?

and in case of not, are you planning for that? if yes, when, please?

**Reply by gmourier** [reply]
Hey @MoustafaElsaghier 👋

It's possible to do that through the API but not in a single API call. That's what is implemented in https://github.com/meilisearch/instant-meilisearch/issues/884#issuecomment-1383951654.

Can you elaborate a bit more on your previous comment? I would like to know more about the following statement

> I'm looking for the disjunctive feature to be in API not a multi-index search

Why do you think it's related to the multi-index search? 

(My 2 cents guess is: the beta of the disjunctive facet search is included with the [instant-meilisearch multi-index search beta](https://github.com/meilisearch/instant-meilisearch/releases/tag/v0.10.1-multi-index-search.0) and it may be unclear from its name that you can do a disjunctive facet computation on a single index). @bidoubiwa Do you confirm it's possible to have a disjunctive facets distribution while searching a single index? 

Please let me know 🙇‍♂️

**Reply by MoustafaElsaghier** [reply]
thank you for your reply
I thought there was another way to handle it in one request so I asked you.

**Reply by gmourier** [reply]
Thank @MoustafaElsaghier

If you are curious about how to do it with the instant-meilisearch beta, we have a [live demo](https://codesandbox.io/s/instant-meilisearch-disjunctive-wi7vqy?file=/src/App.js). As you can see, it can be done for a single index!

**Reply by bidoubiwa** [reply]
It is possible! Thanks for your explainations @gmourier  :)

**Reply by MoustafaElsaghier** [reply]
i'm doing it through API for android cuz couldn't find instant search with V30

**Comment by gmourier** [comment]
Hey everyone 👋

Update on this if you haven't seen the API prototype to solve this need.

For v1.1, we will release the `multi-search` endpoint allowing to send an array of requests in a single HTTP query.

You can find the prototype [here](https://github.com/meilisearch/product/discussions/615) with a section dedicated to the disjunctive facets distribution

Although the proposed solution does not allow natively computing the facets distribution in a single search request, we hope it will be a good step forward in solving this need.

Instant-meilisearch will do this natively using that endpoint; if you don't use instant-meilisearch, you have to do it on your side using this endpoint.

**Comment by gmourier** [comment]
Hey folks 👋

Version [v1.1](https://github.com/meilisearch/meilisearch/releases/tag/v1.1.0) is out and features the `/multi-search` endpoint, allowing you to implement a behavior enabling end-users to continue selecting facet values to refine their search even when a value has already been selected for a facet.

Instant-meilisearch uses this endpoint to implement a "disjunctive facets distribution" natively; we hope to release a how-to guide for people who don't use instant-meilisearch.

📚 https://docs.meilisearch.com/reference/api/multi_search.html#post-route
